### PR TITLE
Implement Reorder Functionality of existing placed order

### PR DESCRIPTION
**Pull request by vishalkushwaha94**
Hi
This is Vishal Kushwaha from Agra. 
I have received an email from Webkul to complete a task of Reorder functionality in bagisto.
I do some changes in code to achieve my goal of reorder functionality so please review my changes.
I have also attache my task pdf that i received in email for reference

[Reorder Feature.pdf](https://github.com/bagisto/bagisto/files/4694140/Reorder.Feature.pdf)

### CODE REVIEW

**Comment by rahulcs0082** [comment]
Hello @vishalkushwaha94 

We highly appreciate your work and effort but there are some changes that you need to do so that we can review your PR for functional purpose.

First of all, we are using multilanguage for this project so you need to write translation for each language instead of English only.

Kindly follow the PSR-2 coding standard. Please go through the given link and update code according to this standard - https://www.php-fig.org/psr/psr-2/  and  https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md

**Comment by vishalkushwaha94** [comment]
I have translated keywords in all language but there are some languages in which i can't find the exact translation of keyword like "Reorder" so in that case it is in English and also updated my code according to PSR-2 Standard.

While adding translation of keyword in Shop Module language file i find some strange things like there is one keyword which is used as title for cancel button, It is not present in every language file. I find it strange that's why i mention it here.

**Comment by shubhwebkul** [comment]
Hi Vishal,

 - Please add support for default theme too.
 - design issue.
issue

File: packages/Webkul/Shop/src/Http/Controllers/OrderController.php [hunk 1]
```diff
@@ -70,7 +83,16 @@ public function view($id)
             abort(404);
         }
 
-        return view($this->_config['view'], compact('order'));
+        foreach ($order->items as $item) {
+            $slugOrPath = strtolower(str_replace(" ","-",$item->name));
+            $product = $this->productRepository->findBySlug($slugOrPath);
+            
+            if(!$product->isSaleable()){
```
Comment (jitendra-webkul) [inline packages/Webkul/Shop/src/Http/Controllers/OrderController.php:90]: use space after ! operator

File: packages/Webkul/Shop/src/Http/Controllers/OrderController.php [hunk 2]
```diff
@@ -70,7 +83,16 @@ public function view($id)
             abort(404);
         }
 
-        return view($this->_config['view'], compact('order'));
+        foreach ($order->items as $item) {
+            $slugOrPath = strtolower(str_replace(" ","-",$item->name));
+            $product = $this->productRepository->findBySlug($slugOrPath);
```
Comment (jitendra-webkul) [inline packages/Webkul/Shop/src/Http/Controllers/OrderController.php:88]: You can directly get product from $item. (Example $item->product)

File: packages/Webkul/Shop/src/Http/Controllers/OrderController.php [hunk 3]
```diff
@@ -61,6 +73,7 @@ public function index()
      */
     public function view($id)
     {
+        $can_reorder = true;
```
Comment (jitendra-webkul) [inline packages/Webkul/Shop/src/Http/Controllers/OrderController.php:76]: Please Camel case here

File: packages/Webkul/Shop/src/Http/Controllers/CartController.php [hunk 4]
```diff
@@ -115,6 +127,27 @@ public function remove($itemId)
...
+    */
+    public function reorder($id)
+    {
+        $order = $this->orderRepository->findOneWhere([
+            'customer_id' => auth()->guard('customer')->user()->id,
+            'id'          => $id,
+        ]);
+
+        foreach ($order->items as $item) {
+            $slugOrPath = strtolower(str_replace(" ","-",$item->name));
```
Comment (jitendra-webkul) [inline packages/Webkul/Shop/src/Http/Controllers/CartController.php:143]: Use space after and before with binary operator

File: packages/Webkul/Shop/src/Http/Controllers/CartController.php [hunk 5]
```diff
@@ -115,6 +127,27 @@ public function remove($itemId)
...
+            $product = $this->productRepository->findBySlug($slugOrPath);
+            Cart::addProduct($product->product_id, ['product_id'=>$product->product_id,'quantity'=>$item->qty_ordered]);
```
Comment (jitendra-webkul) [inline packages/Webkul/Shop/src/Http/Controllers/CartController.php:145]: Use space after and before with binary operator

File: packages/Webkul/Shop/src/Http/Controllers/CartController.php [hunk 6]
```diff
@@ -24,24 +25,35 @@ class CartController extends Controller
...
      */
     public function __construct(
         WishlistRepository $wishlistRepository,
-        ProductRepository $productRepository
+        ProductRepository $productRepository,
+        OrderRepository $orderRepository
     )
     {
-        $this->middleware('customer')->only(['moveToWishlist']);
+        $this->middleware('customer')->only(['moveToWishlist','reorder']);
```
Comment (jitendra-webkul) [inline packages/Webkul/Shop/src/Http/Controllers/CartController.php:49]: Use space after and before with binary operator

File: packages/Webkul/Velocity/src/Resources/views/shop/customers/account/orders/view.blade.php [hunk 7]
```diff
@@ -28,11 +28,26 @@
...
                 @if ($order->canCancel())
                     <span class="account-action">
-                        <a href="{{ route('customer.orders.cancel', $order->id) }}" class="theme-btn light unset pull-right" v-alert:message="'{{ __('shop::app.customer.account.order.view.cancel-confirm-msg') }}'" style="float: right">
+                        <a href="{{ route('customer.orders.cancel', $order->id) }}" class="theme-btn light unset pull-right" v-alert:message="'{{ __('shop::app.customer.account.order.view.cancel-confirm-msg') }}'" style="float: right; margin-left: 8px;">
                             {{ __('shop::app.customer.account.order.view.cancel-btn-title') }}
                         </a>
                     </span>
                 @endif
+                @if (core()->getConfigData('customer.settings.reorder.product_reorder'))
+                    @if($can_reorder)
```
Comment (jitendra-webkul) [inline packages/Webkul/Velocity/src/Resources/views/shop/customers/account/orders/view.blade.php:37]: Use space after if and Use camel case for $can_reorder

File: packages/Webkul/Shop/src/Http/Controllers/OrderController.php (hunk 1 above)
Comment (vishalkushwaha94) [reply packages/Webkul/Shop/src/Http/Controllers/OrderController.php]: done

File: packages/Webkul/Shop/src/Http/Controllers/OrderController.php (hunk 2 above)
Comment (vishalkushwaha94) [reply packages/Webkul/Shop/src/Http/Controllers/OrderController.php]: Thanks for your tip it help me to reduce no of line in my code.

File: packages/Webkul/Shop/src/Http/Controllers/OrderController.php (hunk 3 above)
Comment (vishalkushwaha94) [reply packages/Webkul/Shop/src/Http/Controllers/OrderController.php]: done

File: packages/Webkul/Shop/src/Http/Controllers/CartController.php (hunk 4 above)
Comment (vishalkushwaha94) [reply packages/Webkul/Shop/src/Http/Controllers/CartController.php]: done

File: packages/Webkul/Shop/src/Http/Controllers/CartController.php (hunk 5 above)
Comment (vishalkushwaha94) [reply packages/Webkul/Shop/src/Http/Controllers/CartController.php]: done

File: packages/Webkul/Shop/src/Http/Controllers/CartController.php (hunk 6 above)
Comment (vishalkushwaha94) [reply packages/Webkul/Shop/src/Http/Controllers/CartController.php]: done

File: packages/Webkul/Velocity/src/Resources/views/shop/customers/account/orders/view.blade.php (hunk 7 above)
Comment (vishalkushwaha94) [reply packages/Webkul/Velocity/src/Resources/views/shop/customers/account/orders/view.blade.php]: done

**Comment by vishalkushwaha94** [comment]
I have tried to resolve all the mentioned issues. Please let me know if there are any remaining.

**Comment by vaishaliwebkul** [comment]
@vishalkushwaha94 
   check reorder functionality with all product type 
i.e; getting an exception when trying to reorder a configurable product

Screenshot (33)

**Comment by vishalkushwaha94** [comment]
@vaishaliwebkul 

I have resolved error in reorder functionality with all product type if there is any one remaining please let me know.

**Comment by vaishaliwebkul** [comment]
Hi @vishalkushwaha94 
It's now fixed for all product type :)

I am getting the error alert when reordering configurable product two times from the same order id

Screenshot (54)

**Comment by vishalkushwaha94** [comment]
@vaishaliwebkul 

I have resolved the error please let me know if there are any remaining.

**Comment by vaishaliwebkul** [comment]
@vishalkushwaha94 
can you please check the reorder for event booking type, as I am getting this error when place reorders for event booking even quantity are available

Screenshot (58)

**Comment by vishalkushwaha94** [comment]
@vaishaliwebkul 
I hope this time i have resolved all issues, but please let me know if there are any remaining.

**Comment by vaishaliwebkul** [comment]
@vishalkushwaha94 
Thanks for the efforts, but the last issue is still not resolved, can you please check again.

**Comment by vishalkushwaha94** [comment]
@vaishaliwebkul 
I have already resolved last issue, please check again and also share the screenshot of error msg because i am not facing any error while adding  event booking.


I have already solved the last issue, please check again and also share the screenshot of the error message as I am not facing any errors  while reorder event booking.

**Comment by vaishaliwebkul** [comment]
Hi @vishalkushwaha94 
I have updated my code with [commit](https://github.com/vishalkushwaha94/bagisto/commit/2cb7db38238351e9ab8361b2d99dd37c9ff65471) and check multiple times for this issue but always showing the same result.
Hope this will help you https://www.loom.com/share/40190127d6f245e9afc93410c3cd11c3

**Comment by vishalkushwaha94** [comment]
Hi @vaishaliwebkul 
Thanks for your response, I have resolved the issue, so please check and let me know if there are any issues remaining

**Comment by vishalkushwaha94** [comment]
Hi @vaishaliwebkul

If you face any kind of error in reorder functionality please let me know i will resolve it as soon as possible.

**Comment by vaishaliwebkul** [comment]
Hi @vishalkushwaha94 
you could also fix one more issue, follow the below steps:

1. place order for the rental booking(daily basis)
2. reorder this order( working fine)
3. now edit the same product for an hourly basis and save
4. reorder the previously order again

Screenshot (74)
Screenshot (75)

**Comment by vishalkushwaha94** [comment]
Hi @vaishaliwebkul
Thanks for your response, but I would like to know if only this issue is left or there are others because if you face any other issues please mention here, this will help me to resolve them once.

**Comment by vaishaliwebkul** [comment]
@vishalkushwaha94 
I have tested this functionality with the possible cases, all issues have been reported here
Finally, the last one also is resolved :)

Thanks.

**Comment by ghermans** [comment]
@vishalkushwaha94 can you please resolve the conflicted file?

**Comment by otheriz** [comment]
Just a quick question..

Is this PR already read the availability of the stock?
what will happen if one or some of the 'reorder' items is have no stock?

Any alert or something?

**Comment by vishal-webkul** [comment]
Reorder option is not available if some of the 'reorder' items is have no stock.

**Comment by ghermans** [comment]
I am closing this PR since there has been no activity since 23 Sep 2020.

**Pull request closed without merging**